### ISSUE TITLE
[Performance] SM100: packed varlen L<64 is 4.4x slower than padded L64 (guarded tail staging vs TMA)

### ISSUE THREAD

**Issue by SakiNana7qi** [QwenLM/FlashQLA#43]
## Summary

On NVIDIA B200 (SM100), `chunk_gated_delta_rule` has a large packed-varlen performance cliff when every sequence is shorter than the 64-token chunk size.

For N=4096 sequences of length 32, forward takes 12.78 ms. Padding every sequence to length 64 with neutral values doubles the physical token count, but forward drops to 2.90 ms. Backward shows the same direction, although the gap is smaller (17.07 ms vs 10.48 ms). The worst case in the length sweep is L63, where forward takes 22.42 ms against 2.91 ms for L64 (7.7x).

The cliff is stable across repeated timing blocks. Boundary measurements and Nsight Compute indicate that incomplete chunks use guarded elementwise Q/K/V/A staging, whereas complete 64-token chunks use TMA staging.

## Environment

- GPU: NVIDIA B200, compute capability 10.0, 180 GB HBM
- NVIDIA driver: 580.95.05
- CUDA: 13.0
- Python: 3.12
- PyTorch: 2.11.0+cu130
- FlashQLA: source checkout at `7c7dfe16416ad21b1d03258189fc8d3b8460ae06`, imported directly from the clean checkout with no local patch
- TileLang: 0.1.9, installed as a Python package in the runtime image
- Workload: B=1, N=4096, Hq=Hk=Hv=12, D=128; BF16 Q/K/V and beta; FP32 log-space decay; packed varlen; no initial state; no final state

## Observed behavior

Each latency below is the median of 12 round-level p50 values. Each round uses 20 warmup calls followed by 100 CUDA Event measurements.

| Phase | packed L32 | neutral-padded L64 | L32 / L64 |
| --- | ---: | ---: | ---: |
| Forward | 12.7765 ms | 2.8989 ms | 4.407x |
| Backward | 17.0664 ms | 10.4817 ms | 1.628x |

The largest coefficient of variation among the round-level p50 values was 0.038%. FlashQLA's auto-CP planner selected `use_cp=false` for both shapes in both forward and backward.

## Correctness of the padding control

For every real L32 sequence, the L64 control appends 32 positions with Q/K/V, log-space decay, beta, and backward `do` set to zero. Only the first 32 positions of every sequence are compared.

The standalone reproducer below performs a full output-tensor comparison, not sampling. It reports:

```text
full output max abs: 0
L32 output nonfinite: 0
padded L64 output nonfinite: 0
```

The same standalone run reproduced the performance cliff at 13.0460 ms for L32 versus 2.9518 ms for padded L64, a 4.4196x ratio. The table above uses the more stable 12-round experiment.

In the larger backward harness, 4,096 evenly spaced elements from each of q/k/v/g/beta gradients also had zero relative L2, zero cosine error, zero max absolute error, and no nonfinite values. These are reported observations rather than precision threshold assertions.

## Boundary and profiling evidence

At fixed N=4096, the performance cliff aligns with the 64-token chunk boundary:

| Sequence length | Forward | Backward |
| ---: | ---: | ---: |
| 31 | 12.803 ms | 17.092 ms |
| 32 | 12.785 ms | 17.074 ms |
| 33 | 13.478 ms | 18.155 ms |
| 63 | 22.421 ms | 24.913 ms |
| 64 | 2.911 ms | 10.518 ms |
| 65 | 5.092 ms | 14.628 ms |

For L32, a warmed Nsight Compute capture attributes 12.287 of 12.996 ms of profiled forward kernel time to `fused_gdr_fwd` (94.54%). The 12.996 ms capture time and the 12.78 ms CUDA Event result above use different measurement modes and are not expected to be identical.

L63 and L64 have the same launch count, 49,152-block grid, 512-thread block, register count, and shared-memory allocation. The material change is the staging path: full 64-token chunks use TMA copies, while incomplete chunks use guarded elementwise Q/K/V/A loads. With N=4096 and 12 heads, every L32 sequence creates a tail-only block.

## Minimal forward reproducer

The call uses only arguments present in the current public `chunk_gated_delta_rule` signature. `auto_cp=True`, `state_v_first=False`, and in-kernel Q/K normalization disabled are all left at their current defaults.

```python
import statistics

import torch
from flash_qla import chunk_gated_delta_rule

N, L, PAD, H, D = 4096, 32, 64, 12, 128
torch.manual_seed(20260901)


def normalize(x):
    x32 = x.float()
    return (x32 / x32.square().sum(-1, keepdim=True).sqrt().clamp_min(1e-6)).to(x.dtype)


q = normalize(torch.randn(1, N * L, H, D, device="cuda", dtype=torch.bfloat16))
k = normalize(torch.randn_like(q))
v = torch.randn_like(q)
g = -torch.rand(1, N * L, H, device="cuda", dtype=torch.float32) * 0.02
beta = torch.rand(1, N * L, H, device="cuda", dtype=torch.bfloat16)


def neutral_pad(x):
    out = torch.zeros((1, N * PAD, *x.shape[2:]), device=x.device, dtype=x.dtype)
    out.view(1, N, PAD, *x.shape[2:])[:, :, :L].copy_(x.view(1, N, L, *x.shape[2:]))
    return out


args32 = (q, k, v, g, beta)
args64 = tuple(neutral_pad(x) for x in args32)
cu32 = torch.arange(N + 1, device="cuda", dtype=torch.int32) * L
cu64 = torch.arange(N + 1, device="cuda", dtype=torch.int32) * PAD


@torch.no_grad()
def run(args, cu):
    return chunk_gated_delta_rule(
        q=args[0],
        k=args[1],
        v=args[2],
        g=args[3],
        beta=args[4],
        initial_state=None,
        output_final_state=False,
        cu_seqlens=cu,
    )[0]


@torch.no_grad()
def time_ms(args, cu, warmup=20, repeat=100):
    for _ in range(warmup):
        run(args, cu)
    torch.cuda.synchronize()
    samples = []
    for _ in range(repeat):
        begin = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        begin.record()
        run(args, cu)
        end.record()
        end.synchronize()
        samples.append(begin.elapsed_time(end))
    return statistics.median(samples)


o32 = run(args32, cu32)
o64 = run(args64, cu64).view(1, N, PAD, H, D)[:, :, :L].reshape_as(o32)
print("full output max abs:", (o32 - o64).abs().max().item())
print("L32 output nonfinite:", (~torch.isfinite(o32)).sum().item())
print("padded L64 output nonfinite:", (~torch.isfinite(o64)).sum().item())
print("L32 ms:", time_ms(args32, cu32))
print("padded L64 ms:", time_ms(args64, cu64))
```

## TileLang #2926 control

TileLang #2926 was considered because it restores `cp.async` eligibility for a guarded global-to-shared copy with an explicit zero-fill else branch. An exact parent/head control changed L32 forward by less than 0.01% and backward by less than 0.5%, without reducing the L32/L64 gap. This does not appear to fix the FlashQLA tail cliff.

<details>
<summary>Exact TileLang parent/head control</summary>

Both TileLang variants were built from source with `pip --no-build-isolation`; the two source trees differ only by PR #2926.

- Parent: `5a9b2a54b9a04831886d9b4dfd7a9ad758cd7ebd`
- Head: `e4ef1462a5441b30bfb44c162bd822ce28cbe098`

| TileLang arm | L32 forward | padded L64 forward | L32 backward | padded L64 backward |
| --- | ---: | ---: | ---: | ---: |
| Parent | 12.7174 ms | 2.8386 ms | 15.9790 ms | 9.6707 ms |
| Parent + #2926 | 12.7172 ms | 2.8386 ms | 15.9523 ms | 9.6277 ms |

</details>

## Expected behavior and possible directions

A shorter physical input should not be more than four times slower than a semantically equivalent padded input containing twice as many tokens.

Possible directions appear to be:

1. keep an asynchronous bulk-copy path for incomplete Q/K/V/A tiles, using safe zero fill if supported; or
2. add a dedicated short-tail kernel or shape-aware fallback so thousands of tail-only sequences do not repeat guarded elementwise staging in the full 512-thread kernel.

Would either direction fit the intended SM100 implementation? I am happy to prepare a PR once maintainers agree on the preferred approach.

Related TileLang context:

- https://github.com/tile-ai/tilelang/issues/2759
- https://github.com/tile-ai/tilelang/pull/2926
